### PR TITLE
PEP 1

### PR DESCRIPTION
**Pull request by c-prather**
This is the minimum viable implementation of PEP 1, #816.  Ideally this would include documentation, but I'm really not sure what to write short of an example.

## PR Summary
See #816 for a description of why this is cool.  tl;dr it takes Parthenon's callback model (e.g., driver calls `Update::FillDerived`, which then calls any package's registered `FillDerived` function) and allows user extensions (e.g., driver calls `MyPackages::ApplyFloors` and that calls a bunch of registered `ApplyFloors` functions depending on which packages are loaded).

See [here](https://github.com/AFD-Illinois/kharma/blob/kharma-next/kharma/kharma_package.hpp) for an example extension package type.

## PR Checklist

- [ ] Code passes cpplint
- [ ] New features are documented.
- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [ ] Code is formatted
- [ ] Changes are summarized in CHANGELOG.md
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [ ] Docs build
- [ ] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

**Comment by c-prather** [comment]
@par-hermes format

File: src/interface/packages.hpp [hunk 1]
```diff
@@ -26,15 +27,61 @@ class Packages_t {
...
+  template <typename T>
+  const std::vector<T*> ListPackagesOfType() const {
+    std::vector<T*> sub_list;
+    for (auto package : packages_) {
+      if (T *cast_package = dynamic_cast<T*>(package.second.get())) {
+        sub_list.append(cast_package);
+      }
+    }
+    return sub_list;
+  }
```
Comment (Yurlungur) [inline src/interface/packages.hpp]: I don't super object to the new dictionary machinery here, but I wonder if a better approach might be to expose a loop structure. For example:
```C++
template <typename T, typename F>
void ForAllPackagesOfType(F &f) const {
  for (auto &[name, pkg] : packages_) {
    if (T *cast_package = dynamic_cast<T*>(pkg.get())) {
      f(name, package);
    }
  }
}
```

File: src/interface/packages.hpp [hunk 2]
```diff
@@ -26,15 +27,61 @@ class Packages_t {
...
+  template <typename T>
+  const Dictionary<T*> AllPackagesOfType() const {
+    Dictionary<T*> sub_dict;
+    for (auto package : packages_) {
+      if (T *cast_package = dynamic_cast<T*>(package.second.get())) {
+        sub_dict[package.first] = cast_package;
+      }
+    }
+    return sub_dict;
+  }
```
Comment (Yurlungur) [inline src/interface/packages.hpp]: Looks like you added this twice?

**Review by Yurlungur — COMMENTED** [review]
This seems pretty minimally invasive, so if it buys you a lot, I'm happy to include it.

File: src/interface/packages.hpp (hunk 1 above)
Comment (c-prather) [reply src/interface/packages.hpp]: I considered that, but I think it would be common/useful to add other elements besides just the function call -- e.g. all my loops right now include extra code that can print the function being called, and I can see them as a good point for adding other debugging or common code.

I also sometimes (often) need control of the order that functions are called, e.g. one "floors" function relies on the results of previous.  Currently I just hack it by pulling out the package I know needs to be first, then looping over the others.  A more general approach might be callback "priorities," but that gets hard to follow reading the code and is more complex to implement so it's not clear we'd want it.

Since only maybe half my callers look anything like that prototype, and I gave up trying to cover the different cases.  I'm happy to provide it for other implementations though.

File: src/interface/packages.hpp (hunk 2 above)
Comment (c-prather) [reply src/interface/packages.hpp]: Yeah git duplicated everything merging and I didn't catch this part.  Fixed.

File: src/interface/packages.hpp (hunk 1 above)
Comment (Yurlungur) [reply src/interface/packages.hpp]: Nah it shouldn't matter much.

**Comment by c-prather** [comment]
@par-hermes format

**Comment by c-prather** [comment]
I've had to additionally mark some `StateDescriptor` members `protected` since I now would like to access them from KHARMA package code.  I think all members should be accessible to subclasses, but if there's anything which should still be `private` I can move it back

**Comment by Yurlungur** [comment]
> I've had to additionally mark some `StateDescriptor` members `protected` since I now would like to access them from KHARMA package code. I think all members should be accessible to subclasses, but if there's anything which should still be `private` I can move it back

I think that's fine

**Review by Yurlungur — APPROVED** [review]

**Review by pgrete — APPROVED** [review]
LGTM

**Comment by Yurlungur** [comment]
@bprather looks like the CI is blocking due to a dependency? Is that right?

**Comment by c-prather** [comment]
There aren't unmet dependencies -- it looks like the check itself is broken (`cannot find file ...`).  Could force-merge this and wait for a fix to the check, or remove the check entirely with a new PR.  Happy to do either (I think I introduced the check in the first place)

EDIT upstream issue: https://github.com/gregsdennis/dependencies-action/issues/24

**Comment by jonahm-LANL** [comment]
You are likely the one who would fix it. So I defer: do you want to remove it or try to debug it?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Ben Prather ***@***.***>
Sent: Wednesday, October 18, 2023 10:04:14 AM
To: parthenon-hpc-lab/parthenon ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [EXTERNAL] Re: [parthenon-hpc-lab/parthenon] PEP 1 (PR #907)


There aren't unmet dependencies -- it looks like the check itself is broken (cannot find file ...). Could force-merge this and wait for a fix to the check, or remove the check entirely with a new PR. Happy to do either (I think I introduced the check in the first place)

**Comment by c-prather** [comment]
Looks like it was fixed upstream already! So, I'm gonna hit go on this.
